### DISCUSSION TITLE
pattern_pass_by_ref

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/labview_book#41] [Announcements]
# pattern_pass_by_ref

传值是符合数据流驱动程序的传参方式，在 LabVIEW 中应该尽量使用这种方式。但是传引用在某些情况下是不可避免的。假如程序要在不同的线程中对同一数据进行操作，就不得不用到传引用。

https://lv.qizhen.xyz/pattern_pass_by_ref

**Comment by whchrome** [comment]
阮老师您好，最近编的程序遇到了图 4.67 所示类似的问题，数据存在一个数组中，多个子vi并行处理数组中的相应元素，并更新相应元素的值。如果传值的话，每个子vi获取的是数组的旧值，更新对应元素后再传给数组，这样会导致后处理完的子vi传出的数组其他元素的值，会覆盖先处理完的子vi更新的对应元素的新值。所以我的问题是，图 4.67中的sub1和sub2是怎样的处理方法，从而避免了前述问题？如果采用图 4.68的方法，那输出的引用还是输入的那个引用吗？如果采用图 4.69 所示的元素同址操作结构，是否可以确保100%避免前述问题？

**Reply by ruanqizhen** [reply]
使用“数据值引用”相关节点和使用“元素同址操作结构”在功能上是完全等效的，它们都可以保证不同线程（不同并行子 VI）中对数据的修改不会发生冲突。
在更底层的实现中，计算机是采用了一个“资源锁”来保证数据安全的：当一个线程需要修改数组中的元素时，它会首先检查这个数组有没有被“锁住”，如果有，那么它是不能修改数组数据的，要等到解锁后才行；如果数组没有被锁住，那么这个线程就会把这个数组“锁”起来，以防止其它线程来修改数据，当自己把数据修改好之后，再对数据解锁，这样其它线程才能继续修改数据。
由于“锁”的存在，保证了任何时间，只能有一个线程对数据做修改，而且每次修改都是在最新数据的基础上做修改的，这样就不会出现某次修改被其它线程的数据覆盖的问题了。

**Reply by whchrome** [reply]
明白了，谢谢解答

**Comment by JustinGZJ** [comment]
阮老师您好，
1.我想问一下队列的经过循环为什么要移位寄存器？

**Reply by ruanqizhen** [reply]
任何数据经过循环都最好使用移位寄存器，因为循环次数可能是0。如果循环次数为0，又没有使用移位寄存器，数据就会丢失。可以参考：[关于几个常见的错误的解释](https://lv.qizhen.xyz/debug_errors#for-%E5%BE%AA%E7%8E%AF%E7%9A%84%E9%9A%A7%E9%81%93)

**Comment by ruanqizhen** [comment]
之前博客相关内容下的留言：

博弈
2011年09月7日 22:45 编辑
在用引用对数据进行改变时候总会出现意想不到的结果，有时候行不通，有时候开始运行时候能行，但关闭程序后再去运行，就有问题啦，是时间上还是什么回事，真还搞不懂，看您的东西，挺棒的，希望能找到一点答案，谢了您呢!!

回复
ruanqizhen
2011年09月15日 09:48 编辑
时有时无的错误，很可能是多线程引起的。LabVIEW自动多线程执行代码，代码执行顺序不确定，可能在某种顺序下会出错。
你试着用Sequence结构等方法把代码顺序固定下来看看能否解决问题。

回复
Pingback： 博客版《我和LabVIEW》目录 « 凡人琐忆 编辑
Qizhen
2009年04月9日 15:18 编辑
因为这是传引用的啊，所以不论在哪里运行子VI，它改动的都是同一块数据。所以并不需要把输出接入哪里。

回复
lihua
2009年04月8日 22:47 编辑
您好，关于队列的那个实例有一点没有弄明白，就是那个set number子VI，它有一个输入，输出引用，但是在主VI里面，它的输出并没有接入get data里面，但是程序运行后，结果却发生的改变，请问这是怎么回事，我是一个新手，请教啦！

回复
Qizhen
2009年03月11日 17:30 编辑
是的，在出栈时，会有复制。所以，如果不是必须传引用，可以直接传递大数组。只要子VI编写合理，可以避免数组数据拷贝。有些拷贝是不可避免的，比如取子数组等。

回复
gufei
2009年03月2日 11:28 编辑
请教大侠几个问题：1. 我想用队列保存一个大数组，想让他内存固定，对他进行处理，虽然是传引用了，但是每次要使用这个数组的还是要出栈，这样不还是肯定要复制了一下嘛？2.传入子VI是不是也是也要复制下数据、、、用labview处理大数据真的很头疼啊

回复
刘伟印
2008年08月22日 11:34 编辑
原来一直不太清楚Refnum是个什么东西，现在明白些了，谢谢！

回复

**Reply by ruanqizhen** [reply]
之前博客下的留言：

Qizhen
2010年06月8日 11:42 编辑
从第三方DLL中把数据拿给LabVIEW用，只能copy一份，没有效率更高的办法。不过，如果你自己写DLL,倒是可以利用LabVIEW提供的一些API，对LabVIEW中的数据直接操作，避免拷贝。后一种方式比较麻烦，我还没有看到过实际的应用。

回复
Michael
2010年06月2日 16:42 编辑
Qizhen Ruan，“实际上，就是在CLN节点中，通过配置参数的类型为U32，把一个指针数据的地址传出来，再传递给其它CLN使用”，您说的这个做法我已在目前的程序中实现，但是我觉得并不高效，只能一个一个数据通过CLN来读，特别是当数据量多时，估计LabView对返回的数又copy了，我想的做法是：在LabView中可否直接访问这段内存，或者更好的情况是LabView可以直接处理这段连续的内存的数据而画出波形图。我在网上搜索了很多，但是没找到LabView可以高效访问内存的做法，还请赐教，谢谢。

回复
Qizhen
2010年06月2日 16:15 编辑
Michael Xiong，对应书中的图片，有范例可以下载，你可以把示例下载下来，仔细看一下。实际上，就是在CLN节点中，通过配置参数的类型为U32，把一个指针数据的地址传出来，再传递给其它CLN使用。

回复
Michael
2010年06月2日 15:00 编辑
您好，我有一本您的“我和LabView” ，在 "4.8.5 借助 C 语言" 这节中提到 ： “ ……LabView传递的只是这数据空间地址值。后续程序若需要用到被引用数据，则可以通过地址值 得到数据的内容 ………… 见图 4.65”， 对图4.65我没看明白到底是如何通过地址取到数据的，我目前开发的一个程序就需要这样的功能，LavView 调用DLL(c++) 返回一个c分配的内存的地址，而后在LabView中通过这个地址取到值。

回复
Qizhen
2010年03月30日 14:12 编辑
我对 LABVIEW FOR ARM 毫不了解，所以帮不了你了。你可以去论坛发贴问问。

回复
峰
2010年03月23日 16:51 编辑
从接触LABVIEW到现在，不到一年时间感受到了LABVIEW的强大，今天有幸看到您的日志，受益很多，非常感谢最近了解到LABVIEW FOR ARM 您公司对这方面的说明比较少，我现有LPC2478ARM，ULINK2，看了好久您公司网站的PORTING，没有成功，加上本来英语也不太好，不知道您能不能提供点这方面的帮助，谢谢

回复
Qizhen
2010年03月22日 15:31 编辑
LabVIEW使用值传递，因为它是自动多线程的程序。而C++，C#不是，所以他们可以使用传引用，但是使用C++，C#编写多线程的程序就比LabVIEW麻烦得多。详细的你可以参看一下我专门讨论传值传引用的文章。如果需要传引用可以使用“New Data Value Reference”和“Delete Data Value Reference”节点，配合“In Place Element Structure”结构使用，能够保证资源互斥。

回复
XuePing
2010年03月19日 10:46 编辑
仔细考虑后，使用队列传递对象引用似乎可以实现互斥同步访问同一类对象实例。最后剩下的问题是，出队列操作会阻塞哪一重循环呢（如果该操作位于多重循环里），是离它最近那个？当然我可以通过试验来验证，但还是希望能听听专家的意见。

回复
XuePing
2010年03月19日 09:43 编辑
在LabView里，类对象的传递也是值传的。在实例里有通过队列来实现引用传递。现在似乎有了新的方法，不错。不过还是觉得LabView里应该默认让类对象为引用传递(现代的语言里基本都是这样做的了，如C#)。我在实际应用中的遇到的问题是，可能并行地去访问我的类对象（我有一个循环不停地访问它以采集数据，另一个循环需要偶尔访问该对象发送控制命令出去），访类封装了对独占设备（如串口）的访问。这问题仍困扰我，因为我对LabView还不是很熟悉。如何在LabView里实现对对象的引用访问，同时还要解决互斥访问I/O对象问题。这个目前有些超出个人能力了，望大师指点迷津，谢谢！

回复
Qizhen
2009年11月17日 09:31 编辑
你这个问题问的太笼统了。全局变量与这个传引用节点最大的区别应该是：使用传引用节点仍然需要使用数据线把所有使用到这个数据的地方连接起来，而全局变量不用。

回复
盼盼
2009年11月15日 23:45 编辑
请问全局变量和这个功能相比能差多少呢？

回复
Qizhen
2009年10月22日 20:48 编辑
网友 says:解答一个网友的问题，暂时存一下请教大侠：LabVIEW2009中的New Data Value Reference和Delete Data Value Reference在实际应用中的好处，最好能举例，谢谢。Qizhen says: LabVIEW 以前传引用很麻烦，现在简单了。比如说你要在两个并行的for loop 中同时修改同一个数值数据。以前直接把U32数据，分别传到两个循环结构里是不行的，因为数据一旦分叉，就是独立两份了。若想在两处修改同一数据，就先用 New Data Value Reference生成数据的引用，把引用传递到不同循环结构里就可以了。这类似于C++ 中，int& data，这样的传参方式。网友 says: 但一个数组（或任意数据类型）通过New Data Value Reference变成引用后传到不同的地方，当不同的地方要处理这个数组中的任一个元素都必须要使用Delete Data Value Reference来还原数据（产生复制，当然是临时的），视乎没有解决太多数据流的缺陷。Qizhen says: 数据流自然的传递数据方式是按照值传递，如果没必要，应该使用值传递，一次传给每一个模块，对数据进行处理，尤其是数组这种数据量很大的类型。如果一定要使用Reference，可以配合In Place Element Structure结构来使用 在In Place Element Structure结构内部把数组解开，这样LabVIEW会尽量使用数组原地址，而不是复制数组。

回复
